demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+    std::ifstream proc_stat("/proc/stat");
+    std::string line;
+    std::smatch match;
+    std::getline(proc_stat, line);
+    while (proc_stat.good())
+    {
+        std::regex core_jiffies("^cpu(\\d+)\\s+"
+            "(\\d+)\\s+" // user
+            "(\\d+)\\s+" // nice
+            "(\\d+)\\s+" // system
+            "(\\d+)\\s+" // idle
+            "(\\d+)\\s+" // iowait
+            "(\\d+)\\s+" // irq
+            "(\\d+)\\s+" // softirq
+            "(\\d+)\\s+" // steal
+            "(\\d+)\\s+" // guest
+            "(\\d+)$");  // guest_nice
+        if (std::regex_match(line, match, core_jiffies))
+        {
+            unsigned long idleInfo = stoul(match[5]) + stoul(match[6]),
+                non_idleinfo = stoul(match[2])
+                    + stoul(match[3])
+                    + stoul(match[4])
+                    + stoul(match[7])
+                    + stoul(match[8])
+                    + stoul(match[9]), // it doesn't handle overflows and overflows of /proc/stat values
+                core_id = stoul(match[1]);
+            idleNonIdleCpuStat[core_id].first = idleInfo;
+            idleNonIdleCpuStat[core_id].second = non_idleinfo;
+        }
+        std::getline(proc_stat, line);
+    }
+    return idleNonIdleCpuStat;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::enableHistory(std::size_t historySize) {
+    if (0 == historySize) {
+        disableHistory();
+    } else {
+        this->historySize = historySize;
+        historyEnabled = true;
+        if (!lastEnabled) {
+            prevIdleNonIdleCpuStat = getIdleNonIdleCpuStat(nCores);
+        }
+    }
+}
+
+bool CpuMonitor::isHistoryEnabled() const {
+    return historyEnabled;
+}
+
+void CpuMonitor::disableHistory() {
+    historyEnabled = false;
+}
+
+void CpuMonitor::enableLast() {
+    lastEnabled = true;
+    if (!historyEnabled) {
+        historySize = 1;
+        prevIdleNonIdleCpuStat = getIdleNonIdleCpuStat(nCores);
+    }
+}
+
+bool CpuMonitor::isLastEnabled() const {
+    return lastEnabled;
+}
+
+void CpuMonitor::disableLast() {
+    lastEnabled = false;
+}
+
+void CpuMonitor::collectData() {
+    // store CPU Utulization in range [0, 1]
+    std::vector<std::pair<unsigned long, unsigned long>> idleNonIdleCpuStat = getIdleNonIdleCpuStat(nCores);
+    std::vector<double> cpuLoad(idleNonIdleCpuStat.size());
+    for (std::size_t i = 0; i < idleNonIdleCpuStat.size(); ++i) {
+        unsigned long idleDiff = idleNonIdleCpuStat[i].first - prevIdleNonIdleCpuStat[i].first;
+        unsigned long nonIdleDiff = idleNonIdleCpuStat[i].second - prevIdleNonIdleCpuStat[i].second;
+        cpuLoad[i] = static_cast<double>(nonIdleDiff) / (idleDiff + nonIdleDiff);
+    }
+    prevIdleNonIdleCpuStat = std::move(idleNonIdleCpuStat);
+
+    if (meanCpuLoad.empty()) {
+        meanCpuLoad = cpuLoad;
+        samplesNumber = 1;
```

Comment:
Instead of this, it would be simpler to initialize `meanCpuLoad` to a vector of zeroes.

Reply:
updated